### PR TITLE
Add SNS to PUT endpoint

### PR DESCRIPTION
**Pull request by humulla**
### *What is the problem we're trying to solve*

The PUT endpoint now triggers SNS events every time an update is made. If the client sends through the same object as what's already in the database then the API would fail with NoChangesException. 

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test the complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Trigger event in Activity History Listener

### CODE REVIEW

File: PatchesAndAreasApi/V1/Infrastructure/CompareValues.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using System;
+using Hackney.Shared.PatchesAndAreas.Domain;
+
+namespace PatchesAndAreasApi.V1.Infrastructure
+{
+    public static class CompareValues
```
Comment (adamtry) [inline PatchesAndAreasApi/V1/Infrastructure/CompareValues.cs:7]: Is this used anywhere? Looks like you used the .AreAllSame method instead

File: PatchesAndAreasApi/V1/Gateways/PatchesGateway.cs [hunk 2]
```diff
@@ -102,12 +104,18 @@ public async Task<PatchesDb> DeleteResponsibilityFromPatch(DeleteResponsibilityF
...
             var patch = await _dynamoDbContext.LoadAsync<PatchesDb>(query.Id).ConfigureAwait(false);
             if (patch == null) return null;
+            OldResponsibleEntities = patch.ResponsibleEntities;
+
             if (ifMatch != patch.VersionNumber)
                 throw new VersionNumberConflictException(ifMatch, patch.VersionNumber);
 
+            if (patch.ResponsibleEntities.Count == responsibleEntitiesRequestObject.Count & patch.ResponsibleEntities.AreAllSame(responsibleEntitiesRequestObject))
+                throw new NoChangesException();
+
```
Comment (LBHCallumM) [inline PatchesAndAreasApi/V1/Gateways/PatchesGateway.cs:114]: Im curious, was there a reason for not going with the entity updater pattern?

Comment (LBHCallumM) [reply PatchesAndAreasApi/V1/Gateways/PatchesGateway.cs:114]: NVM. I can see its a PUT endpoint now.

**Review by LBHCallumM — APPROVED** [review]
LGTM. This is quite a big PR, so I might have missed something. 

Next time, I would recommend splitting up the PR into smaller chunks. For example, adding SNS and changing the endpoint to a PUT/gateway changes could be separate PRS. It may add a bit more work on your end, but it makes the whole reviewing process much easier.

File: PatchesAndAreasApi/V1/Infrastructure/NoChangesException.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+using System;
+
+namespace PatchesAndAreasApi.V1.Infrastructure
+{
+    public class NoChangesException : Exception
```
[the remark this answers is not in the document]
Comment (humulla) [reply PatchesAndAreasApi/V1/Infrastructure/NoChangesException.cs:5]: Exceptions are normally kept under infrastructure in our repos

File: PatchesAndAreasApi/V1/Infrastructure/CompareValues.cs (hunk 1 above)
Comment (adamtry) [reply PatchesAndAreasApi/V1/Infrastructure/CompareValues.cs:7]: Oh, I didn't know you could use a method call for a different class like that in C#, very cool! Naming is fine.

**Review by adamtry — APPROVED** [review]

**Pull request merged**